presto-main/src/main/java/com/facebook/presto/execution/SqlQueryManager.java

```diff
@@ -350,8 +356,26 @@ public static Statement unwrapExecuteStatement(Statement statement, SqlParser sq
             return statement;
         }
 
-        String sql = session.getPreparedStatementFromExecute((Execute) statement);
-        return sqlParser.createStatement(sql);
+        Execute execute = (Execute) statement;
+        String sql = session.getPreparedStatementFromExecute(execute);
+        Statement unwrapped = sqlParser.createStatement(sql);
+
+        // validate that we have the right number of parameters
+        validateParameters(unwrapped, execute.getParameters());
+
+        // replace Parameter expressions with the supplied values.
+        ParameterRewriter parameterRewriter = new ParameterRewriter(execute.getParameters());
+        AstExpressionRewriter expressionRewriter = new AstExpressionRewriter(parameterRewriter);
+        return (Statement) expressionRewriter.process(unwrapped, null);
+    }
+
+    private static void validateParameters(Node node, List<Literal> parameterValues)
+    {
+        ParameterCollector collector = new ParameterCollector();
+        collector.process(node, null);
+        if (parameterValues.size() != collector.getParameterCount()) {
+            throw new SemanticException(INVALID_PARAMETER_USAGE, node, format("Incorrect number of parameters: expected %d but found %d", collector.getParameterCount(), parameterValues.size()));
```

Comment:
Also, always `%s` instead of a plain `%d` since it does the same thing but is safer (won't throw if the argument isn't a number)